### ISSUE TITLE
Course Homepage: Artwork aligned to top of screen

### ISSUE THREAD

**Issue by luke-freeman** [razeware/emitron-iOS#521]
In the original design the plan was to have the artwork sit flush with the top of the screen and have a transparent navigation bar. However due to some issues with SwiftUI the navigation bar is white/black and pushes the artwork down. 

### Expected behavior

Artwork sits flush to top of screen

<img width="377" alt="Screenshot 2020-11-06 at 13 07 31" src="https://user-images.githubusercontent.com/5320361/98364558-48b15680-2031-11eb-90e5-2de2325006c9.png">

### Actual behavior

White/Dark navigation bar then artwork

<img width="414" alt="Screenshot 2020-11-06 at 13 10 31" src="https://user-images.githubusercontent.com/5320361/98364660-78f8f500-2031-11eb-9608-f053727d988d.png">

### Environment

* iOS Version: 14
* raywenderlich.com App Version: 1.0.5
* Device: iPhone 11

**Comment by bladebunny** [comment]
Added PR https://github.com/razeware/emitron-iOS/pull/560 to support this feature.  Had to change the link text in the navbar to use .titletext, for now, because the default .accent color does not have enough contrast when displayed over the artwork.  The new NavbarStyle modifier can take a different color if so desired.

**Comment by bladebunny** [comment]
Here's what this change look like setting the navbar transparent. Note that it would also be transparent when scrolled up (over list content).  If this is problematic, it will probably require a solution that hides the scrollbar then positions floating back button in the same position - and finally possible adds a background under that custom back button that becomes opaque as the user scrolls up.  Doable but difficult with SwiftUI.

**Comment by bladebunny** [comment]
[
![Uploading Simulator Screen Shot - iPhone 11 Pro - 2021-03-22 at 16.55.30.png…]()
](url)

**Comment by bladebunny** [comment]
![Simulator Screen Shot - iPhone 11 Pro - 2021-03-22 at 16 55 30](https://user-images.githubusercontent.com/1473399/112073048-74d8fa00-8b30-11eb-8b53-3f68c62b1072.png)

**Comment by bladebunny** [comment]
![Simulator Screen Shot - iPhone 11 Pro - 2021-03-22 at 16 55 36](https://user-images.githubusercontent.com/1473399/112073027-67237480-8b30-11eb-9e78-dccae5b81945.png)

**Comment by luke-freeman** [comment]
@bladebunny Can you add a background on scroll?

Come on ARSENAL!!! :)

**Comment by bladebunny** [comment]
I'll have to research that.  Alternatively, a similar tactic using UIKit would have been to get a reference to the navbar and set it initially transparent, but then make progressively more opaque on scroll up to the navbar height.  I'll have to research how to achieve this on SwiftUI if possible.  This PR should not be merged til then.

**Comment by bladebunny** [comment]
I've looked into this a little and here's what I think so far.

I'm going to go out on a limb here and say this is going to be very difficult to achieve with the current implementation, though I am sure there are those who would disagree.

Here's what's needed to implement this feature:
- Height of the "header" card content that I am displaying to up to the top edge (not readily available given the current design but is gettable)
- Starting scroll positions incl. any insets and offsets
-  Notification of scroll events so I can progressively modify the navbar transparency based on scroll/content offset
- Obviously access to the navbar

Many of these things in themselves are very difficult in SwiftUI because SwiftUI abstracts away a lot of the underlying details (though e.g. using anonymous types like some View - and/or doesn't provide access at all).

Currently in Emitron, these views use a **List** to display items.  **List** incorporates a scroll view but is not a scroll view itself and does not expose its scroll view or scroll behaviour.  It is possible to do something like attach a drag gesture recognizer to this, and you will see lots of examples of people doing that, but that is a poor solution because you still need many of the elements discussed above AND most people don't scroll by dragging.  Rather, they will "flick" the list up and down.  Of course, it may be possible to attach additional recognizers but then you have the issue of trying to correlate and track the offset based on a combination of gestures -- ie say they drag a little and flick a little - without knowing the initial state of the offset (because you have no access to the underlying scroll view).

It would be possible to create a solution capable of tracking scroll position if the base view were a scroll view itself.  (e.g. see https://fivestars.blog/swiftui/scrollview-offset.html) -- however that would mean losing some of the features of the SwiftUI List -- like sections -- which Emitron uses.

It's also possible to wrap a UIKit scroll view using the ViewRepresentable protocol and expose the scrollview events and methods needed to track this.  So, much more doable if the underlying container is a scrollview.

SwiftUI 14.0 includes a new reader object called ScrollViewReader which apparently also works with Lists -- but I am not sure if it would help with this feature because it primarily provides a method that allows scrolling to a specific view in a List or ScrollView -- like UIKit TableView and CollectionViews can do.  I haven't seen that it provides a mechanism to receive scroll updates as this feature requires.

There's also then the issue of complexity and maintenance.  How many custom behaviours do you want to add to implement this feature and how hard is it going to be to maintain this going forward?  I love building custom UI but have had the experience enough times in the past of building unique UI that then became hard to maintain that I am a bit gun shy about doing so these days.

TL;DR - The best chance of achieving this feature I can see today is to convert the underlying container to a scroll view which will expose additional options for solving various of the requirements above.  That would mean building custom solutions for some features, for example, List Sections, that List can do but ScrollView cannot. But it is doable.

**Comment by sammyd** [comment]
@bladebunny Thanks for this detailed research—it's really impressive.

One of the purposes of this app is to provide a reasonably sane code base that can be used for teaching, and a reasonable reference for how to make an app. As such, I'm not overly comfortable with coming up with workarounds such as re-implementing system controls. This makes it less useful as a teaching aid, and increases the maintenance cost. Since we don't (as a company) have a huge amount of capacity to develop emitron right now, I'm keen on keeping it as maintainable as possible.

I am open to making the switch to iOS 14 if that would provide enough benefit—it seems like there is some here, but not a definitive solution.

Although in the original design, it's ok for us to just say "that's not currently possible in SwiftUI".

So I think it's probably acceptable to park this issue for now—leaving it open for possible revisiting post-iOS 15.

Thanks again for the effort and investigation you put into this issue!

**Comment by bladebunny** [comment]
Sounds good.  I think that's the right call FWIW.  Thx! I'm going to close the existing PR as well as it only partially solves the problem and leaves other issues - primarily the back nav now being transparent over the list content.

**Comment by VegetarianZombie** [comment]
Closing for now.
